### PR TITLE
[Tx flow] feat: address input redesign

### PR DESCRIPTION
**Pull request by DiogoSoaress**
## What it solves

Resolves https://github.com/safe-global/safe-wallet-web/issues/2099

## How this PR fixes it
- Implements the Address component according to the new design
- Uses the modified components in the `CreateTokenTransfer`

## How to test:
1. Navigate to the Send tokens flow
2. Observe the new component design in the address inputs
  - The caret only shows if there is any option in the suggested list
  - If you are inputing an address that is not contained in your AB, it is suggested to add it

## Screenshots
Screenshot 2023-06-15 at 15 32 46

Screenshot 2023-06-15 at 15 33 09

Screenshot 2023-06-15 at 17 30 11

Screenshot 2023-06-15 at 17 30 16

## Checklist
* [ ] I've tested the branch on mobile 📱
* [ ] I've documented how it affects the analytics (if at all) 📊
* [ ] I've written a unit/e2e test for it (if applicable) 🧑‍💻

### CODE REVIEW

File: src/components/common/AddressBookInput/index.tsx [hunk 1]
```diff
@@ -25,6 +25,7 @@ const AddressBookInput = ({ name, ...props }: AddressInputProps): ReactElement =
 
   return (
     <Autocomplete
+      disableClearable
```
Comment (DiogoSoaress) [inline src/components/common/AddressBookInput/index.tsx:28]: I've used this prop to collapse the padding in the end of the input

**Comment by katspaugh** [comment]
Another thing that I noticed is that the input jumps when selecting an address. Presumably because of the avatar. I would make the avatar smaller to fit into the input.

File: src/components/common/AddressBookInput/index.tsx [hunk 2]
```diff
@@ -13,37 +15,67 @@ const abFilterOptions = createFilterOptions({
...
-        },
-      }}
-      renderOption={(props, option) => (
-        <Typography component="li" variant="body2" {...props}>
-          <EthHashInfo address={option.label} name={option.name} shortAddress={false} />
+    <>
+      <Autocomplete
+        open={open}
+        onOpen={() => setOpen(true)}
+        onClose={() => setOpen(false)}
```
Comment (iamacook) [inline src/components/common/AddressBookInput/index.tsx:43]: Why does this now need to be controlled?

File: src/components/common/AddressInput/index.tsx [hunk 3]
```diff
@@ -55,73 +76,86 @@ const AddressInput = ({ name, validate, required = true, deps, ...props }: Addre
...
-        <Grid item>
+  const endAdornment = (
+    <InputAdornment position="end">
+      {resolving || isValidating ? (
+        <CircularProgress size={20} />
+      ) : !props.disabled ? (
+        <>
+          <IconButton onClick={() => setOpen(true)}>
+            <SvgIcon component={SaveAddressIcon} inheritViewBox fontSize="small" color="primary" />
+          </IconButton>
```
Comment (iamacook) [inline src/components/common/AddressInput/index.tsx:87]: Nit: I would not show this if the address is already present in the address book.

File: src/components/common/AddressInputReadOnly/index.tsx [hunk 4]
```diff
@@ -0,0 +1,44 @@
+import { useState, type ReactElement } from 'react'
+import { IconButton, InputAdornment, InputLabel, OutlinedInput, SvgIcon, Typography } from '@mui/material'
+import EthHashInfo from '@/components/common/EthHashInfo'
+import SaveAddressIcon from '@/public/images/common/save-address.svg'
+import css from './styles.module.css'
+import EntryDialog from '@/components/address-book/EntryDialog'
+import useAddressBook from '@/hooks/useAddressBook'
+
+const AddressInputReadOnly = ({ label, address }: { label: string; address: string }): ReactElement => {
```
Comment (iamacook) [inline src/components/common/AddressInputReadOnly/index.tsx:9]: A lot of this code is similar to the `AddressBookInput` where we have a read only flag. Can we use that instead?

File: src/components/common/AddressBookInput/index.tsx [hunk 5]
```diff
@@ -13,37 +15,67 @@ const abFilterOptions = createFilterOptions({
...
+      />
+      {canAdd ? (
+        <Typography variant="body2" className={css.unknownAddress}>
+          <SvgIcon component={InfoIcon} fontSize="small" />
+          <span>You are sending tokens to an unknown address. Add it to address book.</span>
         </Typography>
-      )}
-      renderInput={(params) => <AddressInput {...params} {...props} name={name} />}
-    />
+      ) : null}
```
Comment (iamacook) [inline src/components/common/AddressBookInput/index.tsx:77]: Nit: this component is not related to sending assets. I would suggest we move this to where it is relevant. What do you think?

**Review by iamacook — COMMENTED** [review]
The design does not truncate the sender address:

image

File: src/components/common/AddressBookInput/index.tsx (hunk 2 above)
Comment (DiogoSoaress) [reply src/components/common/AddressBookInput/index.tsx:43]: So I needed to make this controlled to pass the handler to the `AddressInput` to control the Autocomplete opening from the endAdornment icon (on top of just clicking the input).

File: src/components/common/AddressInput/index.tsx (hunk 3 above)
Comment (DiogoSoaress) [reply src/components/common/AddressInput/index.tsx:87]: If the address is present in the AB, it renders the `AddressInputReadOnly` and we only show the button there if the address is not contained in the AB.

Comment (iamacook) [reply src/components/common/AddressInput/index.tsx:87]: I see. It probably makes more sense if we use the read only flag and don't keep the `AddressInputReadOnly` component then.

File: src/components/common/AddressBookInput/index.tsx (hunk 5 above)
Comment (DiogoSoaress) [reply src/components/common/AddressBookInput/index.tsx:77]: Given this is the AddressBookInput I think it makes sense to give this information. I would phrase it differently to be more generic wdyt? cc @liliiaorlenko 
> You've input an unknown address. Add it to your address book

File: src/components/common/AddressInputReadOnly/index.tsx (hunk 4 above)
Comment (DiogoSoaress) [reply src/components/common/AddressInputReadOnly/index.tsx:9]: They seem too different to me to abstract them. While `AddressBookInput` contains the Autocomplete component and deals mostly with the shown options, the `AddressInputReadOnly` is more dumb and it only displays the Address in the redesign style.

**Review by katspaugh — APPROVED** [review]
👍

File: src/components/common/AddressBookInput/index.tsx [hunk 6]
```diff
@@ -13,37 +15,67 @@ const abFilterOptions = createFilterOptions({
...
+            name={name}
+            onOpenListClick={hasVisibleOptions ? handleOpenAutocomplete : undefined}
+            isAutocompleteOpen={open}
+          />
+        )}
+      />
+      {canAdd ? (
+        <Typography variant="body2" className={css.unknownAddress}>
+          <SvgIcon component={InfoIcon} fontSize="small" />
+          <span>You&apos;ve input an unknown address. Add it to your address book.</span>
```
Comment (iamacook) [inline src/components/common/AddressBookInput/index.tsx:75]: Do you think we should only show this when it's a valid address?

image

I would also suggest we tweak the wording a bit and have it open the modal when clicking on "adding it to your address book".

```suggestion
          <span>This is an unknown address. Consider adding it to your address book.</span>
```

**Review by iamacook — CHANGES_REQUESTED** [review]

**Review by iamacook — APPROVED** [review]
